dask/array/core.py

```diff
@@ -2006,6 +2006,79 @@ def bincount(x, weights=None, minlength=None):
 
     return Array(dsk, name, chunks, dtype)
 
+def histogram(a, bins=None, range=None, normed=False, weights=None, density=None):
+    """
+    Blocked variant of numpy.histogram.
+    
+    Follows the signature of numpy.histogram exactly with the following
+    exceptions:
+    
+    - either the ``bins`` or ``range`` argument is required as computing 
+      ``min`` and ``max`` over blocked arrays is an expensive operation 
+      that must be performed explicitly.
+    
+    - ``weights`` must be a dask.array.Array with the same block structure
+       as ``a``.
+    
+    Original signature follows below.
+    """ + np.histogram.__doc__
+    if bins is None or (range is None and bins is None):
+        raise ValueError('dask.array.histogram requires either bins '
+                         'or bins and range to be defined.')
+    
+    if weights is not None and weights.chunks != a.chunks:
+        raise ValueError('Input array and weights must have the same '
+                         'chunked structure')
+    
+    if not np.iterable(bins):
+        mn, mx = range
+        if mn == mx:
+            mn -= 0.5
+            mx += 0.5
+        
+        bins = np.linspace(mn, mx, bins + 1, endpoint=True)
+    
+    nchunks = len(list(core.flatten(a._keys())))
+    chunks = ((1,) * nchunks, (len(bins) - 1,))
+    
+    name1 = 'histogram-sum' + next(tokens)
+    
+    
+    # Map the histogram to all bins
+    def block_hist(x, weights=None):
+        return np.histogram(x, bins, weights=weights)[0][np.newaxis]
+    
+    if weights is None:
+        dsk = dict(((name1, i, 0), (block_hist, k))
+                    for i, k in enumerate(core.flatten(a._keys())))
+        dtype = 'i8'
```

Comment:
The return type is different with/without weights. Maybe we should just mock np.histogram?

https://github.com/numpy/numpy/blob/v1.9.1/numpy/lib/function_base.py#L199-L203